File: Week12_LinkedList/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        ans = ListNode(-1)
        excess = 0
        ptrAns = ans
        
        while l1 or l2 or excess != 0:
            # Ternary operators can be used to write this
            l1val = 0
            l2val = 0
            if l1:
                l1val = l1.val
            if l2:
                l2val = l2.val
            
            total = l1val + l2val + excess
            excess = total // 10
            ptrAns.next = ListNode(total % 10)
            
            # Please remember to change pointer reference!
            ptrAns = ptrAns.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        return ans.next

File: Week12_LinkedList/add-two-numbers.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode ans = new ListNode(-1);
        int excess = 0;
        ListNode ptrAns = ans;

        while (l1 != null || l2 != null || excess != 0) {
            int l1val = l1 == null ? 0 : l1.val;
            int l2val = l2 == null ? 0 : l2.val;

            int total = l1val + l2val + excess;
            excess = total / 10;
            ptrAns.next = new ListNode(total % 10);

            ptrAns = ptrAns.next;
            l1 = l1 == null ? null : l1.next;
            l2 = l2 == null ? null : l2.next;
        }

        return ans.next;
    }
}